src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/IDE/VSPackageRestoreManager.cs

```diff
@@ -43,6 +45,10 @@ private void OnSolutionOpenedOrClosed(object sender, EventArgs e)
                     // We need to do the check even on Solution Closed because, let's say if the yellow Update bar
                     // is showing and the user closes the solution; in that case, we want to hide the Update bar.
                     var solutionDirectory = SolutionManager.SolutionDirectory;
+
+                    // go off the UI thread to raise missing packages event
+                    await TaskScheduler.Default;
```

Comment:
Looking at it more closely, this will not make any difference, since this event `OnSolutionOpenedOrClosed` is being triggered on UI thread and then it's using `JTF.Run` which will synchronously execute that async code by blocking calling thread so even though you moved to BG thread, UI thread is still blocked for that operation to be completed. It will even become worse when some where down the line it needs to move back to UI thread which is not part of same JTF context, in that case it will deadlock. And this could be the reason, we reverted earlier code change.

Reply:
Ah, good point. 

Given that this is a synchronous call on the UI thread, I'm not sure there's any thread manipulation we can do to mitigate for this. 

Does that sound right? 

The solution would be probably be avoiding the pattern altogether or trying to avoid heavy work like this in the event handler.  Both of those are a more involved refactoring.

Reply:
yes, any solution here has to be more involved but your best options are either to reduce work being done as part of this call or make it async (use `JTF.RunAsync` instead of `JTF.Run`) if possible.

Reply:
I guess that would help. 
I don't see anything being done during the `RaisePackagesMissingEventForSolutionAsync` that would have issues. It's basically making a call to every PC based project to get the packages and check if they are on disk. If necessary update the restore bar with it. 

The updating of the restore bar is on the UI thread ofc. 

So the flow would need be something like: 

OnSolutionOpenedOrClosed()
// already on the UI thread. 
JTF.RunAsync
   // SwitchToMainThreadAsync 
   // do UI thread stuff. 
  // switch to background thread.  
  RaisePackagesMissingEventForSolutionAsync  
       // do lots of stuff on the background thread. Check if the packages are installed or not. 
      // SwitchToMainThreadAsync 


There is an additional main thread switch here, but it does not seem like too much

I will do some tests to see if I can repro the original issue and see if my change helps that, but does that sound reasonable?

Reply:
Update the fix a little bit. 
The same code path is used during SolutionOpen and SolutionClose event. 

I have concerns changing having an unawaited SolutionClose task, that might coincide with VS being shutdown. We don't really have a delay there so didn't want to risk it so I left the synchrous path unaltered.